pkg/distributor/distributor.go

```diff
@@ -1290,6 +1296,23 @@ func (d *Distributor) cleanupAfterPushFinished(rs *requestState) {
 	}
 }
 
+// checkStartedMiddleware blocks until the underlying distributor service is started.
+func (d *Distributor) checkStartedMiddleware(next PushFunc) PushFunc {
+	return func(ctx context.Context, pushReq *Request) error {
+		next, maybeCleanup := NextOrCleanup(next, pushReq)
+		defer maybeCleanup()
+
+		if d.isServiceStarted.Load() {
```

Comment:
We already have `services.Service.State()` implemented by the distributor struct, would it make sense to use that instead?

Reply:
Yep, done in https://github.com/grafana/mimir/pull/9317/commits/aecfd7321693ace25b39a595ba80fb3a62d4fd41